server/lib/nl/scraper.py

```diff
@@ -205,19 +217,59 @@ def _to_svg(svg):
   return 'data:image/svg+xml;base64,' + svg_b64
 
 
+def _wait_for_svgs_drawn(driver):
+  """Wait (max _MAX_SVG_WAIT_TRIES * 0.5 seconds) for all svgs on a subject page
+  to finish drawing"""
+  svgs_loaded = False
+  num_tries = 0
+  while not svgs_loaded and num_tries < _MAX_SVG_WAIT_TRIES:
+    subject_page = driver.find_element(By.ID, 'subject-page-main-pane')
+    has_incomplete_svg = False
+    for block in subject_page.find_elements(By.CLASS_NAME, 'block'):
+      # Check each chart in the block for any incomplete svgs
+      for chart_container in block.find_elements(By.CLASS_NAME,
+                                                 'chart-container'):
+        svg_tag_to_check = ""
+        for chart_type_class in chart_container.get_attribute('class').split():
+          if chart_type_class in _CHART_TYPE_TO_DRAWN_SVG_TAG:
+            svg_tag_to_check = _CHART_TYPE_TO_DRAWN_SVG_TAG[chart_type_class]
+            break
+        # If there is no svg tag to check for, consider it complete
+        if not svg_tag_to_check:
+          continue
+        drawn_objects = chart_container.find_elements(By.TAG_NAME,
+                                                      svg_tag_to_check)
+        # If there is an svg tag to check for, but it is not found, consider the
+        # chart incomplete
+        if not drawn_objects:
+          has_incomplete_svg = True
+          break
+      # If any block finds an incomplete chart, break out of the loop to wait
+      # and re-check the charts
+      if has_incomplete_svg:
+        break
+    num_tries += 1
+    svgs_loaded = not has_incomplete_svg
+    if not svgs_loaded:
+      time.sleep(0.5)
+
+
 def scrape(query, driver):
   if query.lower() in _OVERRIDE_CHART_MAP:
     return _OVERRIDE_CHART_MAP[query.lower()]
 
   logging.info(f'Scraping: {query}')
   driver.get(_URL + query)
 
-  # Wait until the test_class_name has loaded.
+  # Wait until the page has loaded.
   wait = WebDriverWait(driver, 30)
-  wait.until(EC.presence_of_element_located((By.CLASS_NAME, 'chart-container')))
   wait.until(
       EC.invisibility_of_element_located((By.CLASS_NAME, 'dot-loading-stage')))
-  time.sleep(5)
+  wait.until(EC.presence_of_element_located((By.CLASS_NAME, 'chart-container')))
```

Comment:
To confirm, do we need the dot-loading-stage and chart-container still?

Reply:
actually, can probably get rid of the dot-loading-stage, but should keep the chart-container